From: Ronghua Zang <ronghua@vmware.com>
Date: Thu, 15 Jul 2010 22:18:47 -0700
Subject: [PATCH] net-next: vmxnet3 fixes [2/5] Interrupt control bitmap
References: bnc#624020
Patch-mainline: 2.6.36-rc2
Git-commit: 6929fe8a37365148228206eea8577b3524afc463

A new bit map 'intrCtrl' is introduced in the DriverShared area. The
driver should update VMXNET3_IC_DISABLE_ALL bit before writing IMR.

Signed-off-by: Ronghua Zang <ronghua@vmware.com>
Signed-off-by: Shreyas Bhatewara <sbhatewara@vmware.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Brandon Philips <bphilips@suse.de>

---
 drivers/net/vmxnet3/vmxnet3_defs.h |    7 ++++++-
 drivers/net/vmxnet3/vmxnet3_drv.c  |    5 +++++
 2 files changed, 11 insertions(+), 1 deletion(-)

Index: linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_defs.h
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/drivers/net/vmxnet3/vmxnet3_defs.h
+++ linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_defs.h
@@ -374,6 +374,9 @@ enum vmxnet3_intr_type {
 /* addition 1 for events */
 #define VMXNET3_MAX_INTRS      25
 
+/* value of intrCtrl */
+#define VMXNET3_IC_DISABLE_ALL  0x1   /* bit 0 */
+
 
 struct Vmxnet3_IntrConf {
 	bool		autoMask;
@@ -381,7 +384,9 @@ struct Vmxnet3_IntrConf {
 	u8		eventIntrIdx;
 	u8		modLevels[VMXNET3_MAX_INTRS];	/* moderation level for
 							 * each intr */
-	u32		reserved[3];
+	__le32		intrCtrl;
+	__le32		reserved[2];
+
 };
 
 /* one bit per VLAN ID, the size is in the units of u32	*/
Index: linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_drv.c
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/drivers/net/vmxnet3/vmxnet3_drv.c
+++ linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_drv.c
@@ -71,6 +71,8 @@ vmxnet3_enable_all_intrs(struct vmxnet3_
 
 	for (i = 0; i < adapter->intr.num_intrs; i++)
 		vmxnet3_enable_intr(adapter, i);
+	adapter->shared->devRead.intrConf.intrCtrl &=
+					cpu_to_le32(~VMXNET3_IC_DISABLE_ALL);
 }
 
 
@@ -79,6 +81,8 @@ vmxnet3_disable_all_intrs(struct vmxnet3
 {
 	int i;
 
+	adapter->shared->devRead.intrConf.intrCtrl |=
+					cpu_to_le32(VMXNET3_IC_DISABLE_ALL);
 	for (i = 0; i < adapter->intr.num_intrs; i++)
 		vmxnet3_disable_intr(adapter, i);
 }
@@ -1677,6 +1681,7 @@ vmxnet3_setup_driver_shared(struct vmxne
 		devRead->intrConf.modLevels[i] = adapter->intr.mod_levels[i];
 
 	devRead->intrConf.eventIntrIdx = adapter->intr.event_intr_idx;
+	devRead->intrConf.intrCtrl |= cpu_to_le32(VMXNET3_IC_DISABLE_ALL);
 
 	/* rx filter settings */
 	devRead->rxFilterConf.rxMode = 0;
